### PR TITLE
Redesign the input API a bit

### PR DESCRIPTION
**Pull request by erikrose**
Redesign the input API a bit to guide the caller toward proper use and to divide state. (See commit message.)

test_inkey_0s_raw_ctrl_c fails but only when running `py.test blessed/tests/test_keyboard.py --tb=short`, not when running the whole suite:

    _________________________________________ test_inkey_0s_raw_ctrl_c __________________________________________
    blessed/tests/test_keyboard.py:424: in test_inkey_0s_raw_ctrl_c
        os.write(master_fd, u'\x03'.encode('latin1'))
    E   OSError: [Errno 5] Input/output error

I'm not sure offhand why. Any ideas?
 
And then test_key_mode_no_kb's raises() appears to not work when running the whole suite. When running just test_keyboard, it's fine.

    ____________________________________________ test_key_mode_no_kb ____________________________________________
    blessed/tests/test_keyboard.py:229: in test_key_mode_no_kb
        child()
    blessed/tests/accessories.py:122: in __call__
        assert exc_output == '', exc_output_msg
    E   AssertionError: Output in child process:
    E   ========================================
    E     File "/Users/grinch/Checkouts/blessings/blessed/tests/accessories.py", line 76, in __call__
    E       self.func(*args, **kwargs)
    E     File "/Users/grinch/Checkouts/blessings/blessed/tests/test_keyboard.py", line 227, in child
    E       with term.key_mode():
    E     File "/usr/local/Cellar/python/2.7.3/Frameworks/Python.framework/Versions/2.7/lib/python2.7/contextlib.py", line 17, in __enter__
    E       return self.gen.next()
    E     File "/Users/grinch/Checkouts/blessings/blessed/terminal.py", line 577, in key_mode
    E       raise NoKeyboard
    E   -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
    E   NoKeyboard
    E   ========================================

But that aside, this is ready for feedback.

What do you think? I'm particularly interested in your opinion on key_mode() raising an immediate NoKeyboard exception if there's no keyboard or tty attached. I don't see much call for setting cbreak or raw mode unless you can also take input (in which case we'd defer the error until key() was called), but you may know something I don't.

### CODE REVIEW

**Comment by erikrose** [comment]
@jquast Here this is, in case you didn't get pinged.

**Comment by jquast** [comment]
This is an identity of NoKeyboard exception due to a forked process, NoKeyboard exception must be imported within the child() function, upcoming PR to address.

I'm surprised you now favor to raise NoKeyboard exception, great lengths were made to allow programs to be piped to less(1) which may use keyboard input. ``./bin/progress_bar.py | less`` now raises exception, for example. Your call anyhow. This completely negates ``BufferedKeyboard._has_tty``, as it would always be False in such case, upcoming PR to remove that, then.

I'm trying to update README about how key_mode returns a callable, and I have to say it certainly makes for a quite a mouth-full to document -- previously, only referring to term.inkey, now it is "the callable returned by the context manager key_mode".

Renaming Keystroke as ``Key`` and naming the callable return ``key`` is also certainly confusing, we now must document this ``Key`` class and the "callable returned by..." phrase, which you named ``key`` -- it is such a common idiom for dictionaries in python, such as exceptions KeyError. This is very difficult to wire a brain to read about without rejecting the built-in concept of key.

**Comment by erikrose** [comment]
> I'm surprised you now favor to raise NoKeyboard exception, great lengths were made to allow programs to be piped to less(1) which may use keyboard input. ./bin/progress_bar.py | less now raises exception, for example. Your call anyhow. This completely negates BufferedKeyboard._has_tty, as it would always be False in such case, upcoming PR to remove that, then.

Eek! You're right. Let me rethink.

I also had second thoughts about renaming Keystroke as soon as I pushed the commit.=

**Comment by erikrose** [comment]
About "key" and "keystroke": I waffled back and forth between those. dict.keys() is plural, but one does talking about "key" in the singlar from time to time. Your point about KeyError pushes me the rest of the way back over to keystroke. I'll make the change.

**Comment by erikrose** [comment]
> great lengths were made to allow programs to be piped to less(1) which may use keyboard input

I gave this some thought. I think there are limits to how much pipeability we can accomplish automatically. I always said that one should examine vars like `does_styling` and explicitly decide, based thereon, whether to draw—well—"frippery" was the word I used. Now, in most cases, we could return '' as if the call to key() timed out. However, if somebody calls the block-forever variant key(0), we're out of luck. Can you think of any way to make it behave on a no-input Terminal that won't confuse and surprise the caller, which had quite reasonably assumed some input? I could hear arguments for key(>0) quietly returning '', but I'm honestly having trouble coming up with a scenario in which piping an interactive program makes sense in principle. Perhaps there might be some batch job that would churn happily along if unmolested, but if you pressed P it would pause.

After typing the above, I could see making key(>0) quietly return '' and key(0) raise NoKeyboard. How does that strike you? (I may change my mind again halfway through implementing; I'm inventing rather than extracting here, always a dangerous proposition. So please do inform my musings with concrete usage scenarios from your BBS.)

**Comment by jquast** [comment]
I implemented the ability to do timed polling w/o keyboard to continue the blessed philosophy of "everything is safe for piping".  I could provide examples, but they're trivial. It's your call, I made great effort in trying to remain compatible with blessings API, you're certainly free to change your mind about any such things. If you haven't seen pull #74, it fixes the ./bin programs so that you may try/test keyboard input, as well as the test cases.

------

Looks like you changed your mind about Key, then? Everybody, and I mean everybody, uses "key" to refer to dictionary values (or a dictionary key, or the first item of a 2-tuple in series, or something to do with using zip or iterators) I just picked out a ~8,000 LOC project laying here, it has 220 uses of the phrase "key" and they all have this meaning.

This would be like naming a class Value() for something that provides "business value", or a function named 'object' that, when called with an argument, returns "An objective opinion". It is hard to document, read, and write without triggering unrelated portions of the brain that cause cross-talk and delay comprehension. But the worst offense of all is search/grep-ability.

------

(As for my bbs, the Terminal is always in character-at-a-time mode and uses the ``inkey()`` method, so the context manager ``with key_mode()`` would be an unnerving complication and introduce over-indentation. blessed will have to remain at least here as well as all other projects already using blessed that would require porting.)

**Comment by jquast** [comment]
I should also not that it is **necessary** to derive and modify getch(), kbhit(), and inkey() in my bbs, another reason I couldn't switch back to blessings with this api change.

**Comment by jquast** [comment]
I'm further concerned about separating state using the context manager.

Scripts that are written with different functions, each accepting input and using this context manager might end up slicing the input stream into parts.

This might not be noticed during local development, but only occur when a program is run over say, ssh, the methods formally known as ``kbhit()`` and ``getch()`` may only receive 2 out of 5 bytes of a multibyte sequence or utf-8 sequence, the inkey() method with a small timeout (say, 0 or 0.01) would return -- and then a function might return to a *different* context manager callable return of ``key_mode()``, which would then receive the 2nd half of a multibyte sequence.  Both would cause garbage to be decoded. Do I make sense?

This fix would be (as I have done in the bin/ example programs) is to explicitly pass the callable returned by the context manager to any delegating functions (sometimes in additional to the terminal instance).  This would have to be carefully documented, as well.

**Comment by erikrose** [comment]
Hey, Jeff. Wow, that's a lot of prose. I'll look at it piece by piece.

First, let's not stack up pull requests on top of other pull requests. That's how we got into such a tangle with git last time. If I were to add commits to this one, #74 would get harder and harder to merge. For now, I propose merging #73 to blessed-integration for expediency (even though it isn't perfect), and #74 on top of it (which I haven't even really looked at yet, but I'm going out on a limb) and, in the future, trying to stick to independent PRs.

**Comment by jquast** [comment]
Well then the PR can be discarded entirely, since it is just documenting and fixing tests for this PR ...

**Comment by erikrose** [comment]
I'm attending a Project Night tonight and will mull over this some more.

**Comment by erikrose** [comment]
How does this strike you? I believe it addresses most of your points and makes porting from blessed mostly a matter of renaming:

```python
class Terminal(object):
    """Input buffer moves back onto Terminal."""

    with keystroke_input():  # Renamed from "keystroke_mode", since none of the other mgrs say "mode".

    def keystroke():
        """Keep original, single-char-returning fallback behavior for when tty is in cooked mode, but raise NoKeyboard when timeout is None, lest the program hang forever. If we wanted to be nice, we could even return '' immediately when there is a timeout but no keyboard."""
    
    def _next_char():
    def _char_is_ready():
        """These 2 methods move back onto Terminal."""
```

P.S. A few points illuminating motivations:

* Part of my motivation for moving the input buffer to BufferedKeyboard before was to implicitly flush the input buffer between `with key_mode` invocations. It doesn't, I now see, since the buffering I was looking at is actually just to catch undecodable partial code points. (Ultimately, I predict we'll want to add a flush method to Terminal to avoid unintended buffering between disparate parts of a program: for example, one that makes the user choose from a menu, then does several minutes of processing, then asks for another menu choice. A mistaken key bounce on the initial choice or a cat walking across the keyboard during the processing phase would result in an unintended input on the second menu.)
* Just to clarify, it was always my intent for callers to pass the key() routine around, but, if we're willing to keep the (still kind of strange, I think) single-char fallback behavior of keystroke(), that is one fewer thing to pass around, as you say.

**Comment by jquast** [comment]
> was to implicitly flush

It is the very difficulty of describing this implicitness in documentation -- the code and tests for managing edge-cases, instructing the user why they need it, or not, and how to work with or around it, Pep 20 rule number 2, "Explicit is better than implicit." I don't know how to address the "be careful with the context manager state, make sure to pass it around, danger if you don't" documentation in a "easy for newcomers quick-start guide". It would have to be addressed early (most people don't read beyond the first paragraph -- until they have trouble, anyway) without being too verbose with details.

All of the other existing context managers (other than fullscreen -- though safe, has unexpected behavior) may extend into multiple scopes without side effects.

> is actually just to catch undecodable partial code points.

Are you saying you don't believe this to be a problem?

This is exactly what I'm tried very hard to prevent.

> Just to clarify, it was always my intent for callers to pass the key() routine around

It is my experience that very few places where ``inkey()`` is used without something else from Terminal, where there is input, there is often output prompting for input.

> fallback behavior for when tty is in cooked mode,

Well, we don't know this to be true at all ... your shells default to canonical input, that's all. blessed previously advertised that it could be intermixed with curses programs -- this advertisement would have to be removed or occluded by a caveat: for keyboard input you **must** use *our* context manager. And you don't have to! it works fine, and actually its pretty darn great for debugging:

```
In [1]: import blessed
In [2]: blessed.Terminal().inkey()
^[[A
Out[2]: KEY_UP

In [7]:
```

> I predict we'll want to add a flush method to Terminal

its a rare need, actually many users expect their input to be queued, there are very, very few examples of programs in a unix terminal that behave otherwise (i can't think of a one).  The only place I can think of this as possible is x86 assembler when reading raw scan codes.

If you're going to navigate menus, you don't want to rapidly chose "1, 2, 3", only to be stuck at menu number 2, it actually causes quite a bit of confusion to the user, especially as they become more familiar.  As you add latency (such as ssh), this becomes particularly rude.

Cases in which I would use it is, "I just pasted an entire wikipedia article into this editor, but an error or exception occurred, and its prompting me to make a decision". Anyway I do supply such a property in x/84, ``session.flush_event('input'))``. I actually don't use it anywhere. You can easily do without such a method, you would just do:

```
while term.inkey(0):
    pass
```

Anyway I'm fine with the renames, even though I really preferred my original names, this battle has gone on far enough, I can see you simply don't like them.

**Comment by erikrose** [comment]
> It is the very difficulty of describing this implicitness in documentation -- the code and tests for managing edge-cases, instructing the user why they need it, or not, and how to work with or around it

It doesn't matter now, but that's the reason I was returning the "inkey" callable from the context manager: you can't possibly screw it up, because there was no Terminal-dwelling inkey() to call outside the mgr's scope. (Using as "as" variable outside the scope might be technically possible but would definitely qualify as weird.

> Anyway I'm fine with the renames, even though I really preferred my original names, this battle has gone on far enough, I can see you simply don't like them.

Great. I will have a go.

**Comment by erikrose** [comment]
What does the underscore at the beginning of `_intr_continue` signify?

**Comment by jquast** [comment]
nothing

**Comment by erikrose** [comment]
py.test is running the tests in a continual loop, failing all the while. How do you make it act normally?

**Comment by jquast** [comment]
This happens when you cause an exception after fork, the sub-process is
also py.test, and wants to continue onto the next test after an
exception is raised so you have a "forked testrunner".

Find and fix the first exception and fix it.

**Comment by erikrose** [comment]
Wow, that's exotic. ``-x`` it is.

**Comment by jquast** [comment]
This is the unittest.TestCase fix:

https://github.com/pexpect/pexpect/pull/152/files

you'd need something similar the as_subprocess decorator, I imagine.

**Comment by erikrose** [comment]
How about that? Your design is basically unchanged. All that's mine is the names.

Incidentally, I'm getting a test failure (on OS X), but I was getting it on blessed-integration as well:

```
% py.test blessed/tests/test_keyboard.py --tb=short 
============================================ test session starts ============================================
platform darwin -- Python 2.7.3 -- py-1.4.26 -- pytest-2.6.4
collected 36 items 

blessed/tests/test_keyboard.py ...............F....................

================================================= FAILURES ==================================================
____________________________________ test_keystroke_0s_raw_input_ctrl_c _____________________________________
blessed/tests/test_keyboard.py:456: in test_keystroke_0s_raw_input_ctrl_c
    os.write(master_fd, u'\x03'.encode('latin1'))
E   OSError: [Errno 5] Input/output error
==================================== 1 failed, 35 passed in 9.22 seconds ====================================
```

**Comment by erikrose** [comment]
Obviously, we'll want to reposition your fixes to the demo programs over that.

**Comment by jquast** [comment]
What is your output for ``./tools/display-sighandlers.py`` ?

travis-ci appears to fail because of a PEP8 check. I will copy the branch to my project so I can run it on my build server farm (osx, linux, solaris).

**Comment by erikrose** [comment]
```
% ./tools/display-sighandlers.py
name       value description
---------------------------------
SIGABRT    6     default(SIG_DFL)
SIGALRM    14    default(SIG_DFL)
SIGBUS     10    default(SIG_DFL)
SIGCHLD    20    default(SIG_DFL)
SIGCONT    19    default(SIG_DFL)
SIGEMT     7     default(SIG_DFL)
SIGFPE     8     default(SIG_DFL)
SIGHUP     1     default(SIG_DFL)
SIGILL     4     default(SIG_DFL)
SIGINFO    29    default(SIG_DFL)
SIGINT     2     <built-in function default_int_handler>
SIGIO      23    default(SIG_DFL)
SIGIOT     6     default(SIG_DFL)
SIGKILL    9     None
SIGPIPE    13    ignored(SIG_IGN)
SIGPROF    27    default(SIG_DFL)
SIGQUIT    3     default(SIG_DFL)
SIGSEGV    11    default(SIG_DFL)
SIGSTOP    17    None
SIGSYS     12    default(SIG_DFL)
SIGTERM    15    default(SIG_DFL)
SIGTRAP    5     default(SIG_DFL)
SIGTSTP    18    default(SIG_DFL)
SIGTTIN    21    default(SIG_DFL)
SIGTTOU    22    default(SIG_DFL)
SIGURG     16    default(SIG_DFL)
SIGUSR1    30    default(SIG_DFL)
SIGUSR2    31    default(SIG_DFL)
SIGVTALRM  26    default(SIG_DFL)
SIGWINCH   28    default(SIG_DFL)
SIGXCPU    24    default(SIG_DFL)
SIGXFSZ    25    ignored(SIG_IGN)
```

**Comment by jquast** [comment]
same here (osx, passing), how about ``./tools/display-terminalinfo.py`` ?

**Comment by erikrose** [comment]
```
% ./tools/display-terminalinfo.py
os.isatty(0) => True
locale.getpreferredencoding() => UTF-8
           pathconf   value
------------------- ---------------------------
       PC_MAX_INPUT   1024
        PC_VDISABLE   255
         PC_SYNC_IO   0
        PC_NAME_MAX   [Errno 22] Invalid argument
       PC_MAX_CANON   1024
         PC_PRIO_IO   0
PC_CHOWN_RESTRICTED   200112
        PC_ASYNC_IO   1
        PC_NO_TRUNC   [Errno 22] Invalid argument
    PC_FILESIZEBITS   [Errno 22] Invalid argument
        PC_LINK_MAX   32767
        PC_PIPE_BUF   512
        PC_PATH_MAX   [Errno 22] Invalid argument

Input Mode   Value   Description
---------- -------   ----------------------------------
    IGNPAR     off   ignore (discard) parity errors
     ICRNL      on   map CR to NL (ala CRMOD)
    BRKINT      on   map BREAK to SIGINTR
    ISTRIP     off   strip 8th bit off chars
      IXON      on   enable output flow control
     INLCR     off   map NL into CR
     IXANY      on   any char will restart after stop
     IUCLC   undef   translate upper case to lower case
     INPCK     off   enable checking of parity errors
    PARMRK     off   mark parity and framing errors
    IGNBRK     off   ignore BREAK condition
     IGNCR     off   ignore CR
   IMAXBEL      on   ring bell on input queue full
     IXOFF     off   enable input flow control

Output Mode   Value   Description
----------- -------   ----------------------------------
      OPOST      on   enable following output processing
      ONLCR      on   map NL to CR-NL (ala CRMOD)
      ONOCR     off   No CR output at column 0
     ONLRET     off   NL performs CR function
      OLCUC   undef   translate lower case to upper case
      OCRNL     off   map CR to NL
     ONOEOT   undef   discard EOT's `^D' on output)
     OXTABS   undef   expand tabs to spaces

Control Mode   Value   Description
------------ -------   -------------------------------
  CCTS_OFLOW   undef   CTS flow control of output
      PARENB     off   parity enable
      CLOCAL     off   ignore modem status lines
  CRTS_IFLOW   undef   RTS flow control of input
      MDMBUF   undef   flow control output via Carrier
      PARODD     off   odd parity, else even
         CS6      on   6 bits
     CRTSCTS     off   same as CCTS_OFLOW
       CSIZE      on   character size mask
         CS8      on   8 bits
       CREAD      on   enable receiver
         CS5     off   5 bits (pseudo)
         CS7      on   7 bits
       HUPCL      on   hang up on last close
      CSTOPB     off   send 2 stop bits

Local Mode   Value   Description
---------- -------   ----------------------------------
     ECHOE      on   visually erase chars
NOKERNINFO   undef   no kernel output from VSTATUS
    TOSTOP     off   stop background jobs from output
    ECHONL     off   echo NL even if ECHO is off
    NOFLSH     off   don't flush after interrupt
    PENDIN      on   XXX retype pending input (state)
      ECHO      on   enable echoing
   ECHOPRT     off   visual erase mode for hardcopy
    IEXTEN      on   enable DISCARD and LNEXT
 ALTWERASE   undef   use alternate WERASE algorithm
   ECHOCTL      on   echo control chars as ^(Char)
    ICANON      on   canonicalize input lines
    FLUSHO     off   output being flushed (state)
    ECHOKE      on   visual erase for line kill
   EXTPROC   undef   external processing
      ISIG      on   enable signals INTR, QUIT, [D]SUSP

        Special line Characters
Index               Name    Value
-----------------   ------- ----------
VLNEXT              LNEXT   '\x16'
VDISCARD            DISCARD '\x0f'
VWERASE             WERASE  '\x17'
VEOL                EOL     _POSIX_VDISABLE
VKILL               KILL    '\x15'
VREPRINT            REPRINT '\x12'
VSUSP               SUSP    '\x1a'
VEOF                EOF     '\x04'
VERASE              ERASE   '\x7f'
VEOL2               EOL2    _POSIX_VDISABLE
VSTATUS             STATUS  undef
VMIN                ---     '\x01'
VQUIT               QUIT    '\x1c'
VSTART              START   '\x11'
VTIME               ---     '\x00'
VSTOP               STOP    '\x13'
VINTR               INTR    '\x03'
VDSUSP              DSUSP   undef

os.ttyname(0) => /dev/ttys000
os.ctermid() => /dev/ttys000
```

**Comment by jquast** [comment]
No difference there, either. Build is passing on OSX, Linux, and Travis-CI with the pep8 fix.

Travis-CI has similar issues and is disabled there. The effect is that ^C is sent to the the "master pty" towards the child process, which is interpreted as a literal KeyboardInterrupt signal, (raw mode is not honored by the pty), so the process and its pipe is broken. but it's hard to say unless I can reproduce and sprinkle in some debugging.

I tested manually, seems to work
```
In [5]: with term.keystroke_input(raw=True):
   ...:     inp = term.keystroke()
   ...:
### (pressed ^C here)
In [6]: inp
Out[6]: u'\x03'
```

https://teamcity-master.pexpect.org/viewQueued.html?itemId=2100
https://travis-ci.org/jquast/blessed/builds/47881484

**Comment by jquast** [comment]
also, pep8 fix is 29181e1fc7147c0e0a1a5a3f137d72f88d86dde2 of the same branch name on my repository (i can't push to yours), its simple anyway:

```
--- a/blessed/tests/test_keyboard.py
+++ b/blessed/tests/test_keyboard.py
@@ -172,7 +172,11 @@ def test_char_is_ready_interrupted_no_continue():


 def test_char_is_ready_interrupted_nonetype_no_continue():
-    "_char_is_ready() may be interrupted when _intr_continue=False with timeout None."
+    """
+    _char_is_ready() may be interrupted.
+
+    When _intr_continue=False with timeout None.
+    """
```

**Comment by jquast** [comment]
also, i'm using bash and ``tox`` to execute tests, if it makes any difference. I manually installed pytest and mock into a virtualenv and used vanilla ``py.test``, i have to clear out the current environment a bit (build/ folder, all *.pyc and __pycache__ files and folders), but it still works.

**Comment by jquast** [comment]
looks like you're missing coverage for statement ``raise NoKeyboard(``, https://coveralls.io/files/436593228#L689

**Comment by erikrose** [comment]
And, inexplicably, my tests now pass locally, 100%. I'll take it.

**Comment by erikrose** [comment]
I move to fix all those PEP 8 issues in one go. None of the docstrings in there wrap to 80 columns, and they all violate PEP 257 with their non-triple quotes.

**Comment by erikrose** [comment]
Oh, the actual test are failing? Blech. Alright, I'll patch up that one pep 8 thing.

**Comment by erikrose** [comment]
Merged to blessed-integration, so you can rebase your bin/ fixes and such if you like and open a PR.

**Comment by jquast** [comment]
So ``keyboard_input`` is still a context manager in the branch blessed-integration... are you really set on having this as a context manager? It really is a bit of a no-go for me, I can't even begin to document the "How you should use the callable returned by the context manager" for a laymen, nevermind the "and here's why you shouldn't do this and that, because of these edge case behaviours", nor testing for them.

**Comment by jquast** [comment]
``raise NoKeyboard`` has no test coverage https://coveralls.io/builds/2015588/source?filename=blessings%2Fterminal.py#L689
